### PR TITLE
Allow specifying intermediate textures for cameras.

### LINKED ISSUE #13011: Allow usage of Rg11b10Float for HDR
## What problem does this solve or what need does it fill?

This potentially gives a huge performance boost on weaker mobile devices.

E.g. in my game, I just changed `ViewTarget::TEXTURE_FORMAT_HDR` to `Rg11b10Float` and fps went from 70 to 100 on my Xiaomi Note 8 Pro (midrange phone from 2019). In SDR the fps was about 110. No other changes were required to make it work.

## What solution would you like?

It could be a feature flag or maybe some kind of more convenient run time option.

This format seems to be pretty inaccurate (https://bartwronski.com/2017/04/02/small-float-formats-r11g11b10f-precision/) so probably shouldn't be used all the time.

### PR DESCRIPTION
**Pull request by Brezak**
# Objective

Currently we only support [`Rgba8UnormSrgb`](https://github.com/bevyengine/bevy/blob/f1db525f143ca88a37b26158a43a978d04a8859a/crates/bevy_render/src/texture/mod.rs#L165) and [`Rgba16Float`](https://github.com/bevyengine/bevy/blob/f1db525f143ca88a37b26158a43a978d04a8859a/crates/bevy_render/src/view/mod.rs#L218) as intermediate texture formats for the camera (chosen between by the `is_hdr` field on the camera struct).
This isn't ideal for low power devices where we might want to use `Rg11b10Float` instead of `Rgba16Float` for faster rendering while still supporting effects like bloom. Closes #13011.

## Solution
TLDR:
- Add the `ViewTargetFormat` enum which describes a subset of [texture formats](https://docs.rs/wgpu/latest/wgpu/enum.TextureFormat.html) which can be used as render attachments.
- Replace the `is_hdr` field with a field describing what `ViewTargetFormat` we want to use for the intermediate texture.
- Change most pipeline keys to take a target format they should render to.

> [!TIP]
> Don't miss the detailed description of this PR that I collapsed because It's a wall of text.
<details>

I've replaced the `is_hdr` flag with a `target_format` field to let the user define what texture format the intermediate render passes should use.

Since not all texture formats can be render attachments, I've chosen to isolate a subset that I know *can* be supported. This subset is the enum `ViewTargetFormat`. I mostly pulled information from this [table](https://www.w3.org/TR/webgpu/#plain-color-formats) in the WebGpu working draft. The precise selection criteria are specified in a [note comment](https://github.com/bevyengine/bevy/blob/280719ebcd40ba7891bcd2e7d06c3febbb739b8c/crates/bevy_render/src/texture/mod.rs#L181-L184) in the enum definition. Because not all of these formats are supported by all GPUs I've also added a `SupportedViewTargetFormats` resource that the user can query to check if a given format is supported.

I've also changed several pipeline keys to contain a target format field instead of an is_hdr flag. This change makes up most of the LOC additions in this PR and is mostly noise.

Finally, I tried to figure out what the is_hdr flag implied other than changing the render target format. I've chosen to split the view target formats into clamped and unclamped. The clampedness(?) of a view target format is determined by whether their values are clamped to [0, 1] in the shader. For clamped formats, the code behaves as it would for situations where the `is_hdr` flag from the extracted view or pipeline key was set to false. Conversely, unclamped formats are handled as if is_hdr was true. I'm concerned this is the wrong distinction to make but I don't know enough about rendering to tell what the correct one should be.

I have questions about the bloom pipeline. Specifically, it's the use of `BLOOM_TEXTURE_FORMAT`. Before this PR the rendering flow would presumably feed a `Rgba16Float` formatted texture into the bloom downsampling pipeline which would output a `Rg11b10Float` formatted texture. Said `Rg11b10Float` would find its way into the final bloom upsampling pipeline which would output a `Rgba16Float` texture. My questions are, is it okay for the pipeline to be fed a `Rg11b10Float` texture from the start (as can happen with this PR), and what format should we use when `Rg11b10Float` is not supported?

</details>

## Concerns

- [ ] Is clampedness the correct concept to use instead of HDR (See the details section or the [comment bellow](https://github.com/bevyengine/bevy/pull/13146#issuecomment-2085980107))
- [ ] How should we handle `BLOOM_TEXTURE_FORMAT` (see details section)

---

## Changelog

TODO

## Migration Guide

TODO

### CODE REVIEW

**Comment by Brezak** [comment]
One major thing that should be discussed is, if we should be using the word unclamped in contexts where hdr had been before.

To explain so you don't have to wade through the changes. The values of `ViewTargetFormat` are split into clamped and unclamped. Clamped formats are as the name implies, clamped to some range of values in the shader (currently [0, 1] for every clamped format we support). Conversely, unclamped formats can attain any floating point value (don't quote me on this, I have no idea how NaNs and infinities are handled). In this pr, view targets with unclamped formats are handled as we used to handle hdr view targets (bloom is supported and tonemapping happens as a separate step of the rendering pipeline) while view targets with clamped formats are handled as we would have handled non-hdr view targets.

**Comment by Brezak** [comment]
> Does this allow avoiding the RG11B10UFLOAT_RENDERABLE requirement by tonemapping into rgba8? That would be nice for me personally.

I'm sorry, but I'm too out of my depth with this PR to answer your question. Most everything I know about Bevy rendering and rendering, in general, I learned while writing this PR. What I can tell you, though it probably won't help, is that the only place I found the engine using the `Rg11b10` texture format was as the intermediate output format of the whole bloom pipeline. This hasn't changed.

**Comment by IceSentry** [comment]
Can you fix the conflicts? The PR direction looks overall good to me, but I'd prefer waiting on the conflicts being fixed before doing a full review.

**Comment by IceSentry** [comment]
Also, would it be possible to split this in 2 PRs? One that add the `ViewTargetFormat` and another that updates the pipeline to remove the `is_hdr`? That way the PR will be a bit easier to follow without all the noise.

File: crates/bevy_render/src/camera/camera.rs [hunk 1]
```diff
@@ -886,6 +893,23 @@ pub fn extract_cameras(
 
             let mut commands = commands.get_or_spawn(entity);
 
+            let target_format = if target_formats.is_supported(camera.target_format) {
```
Comment (tychedelia) [inline crates/bevy_render/src/camera/camera.rs:896]: Does it really make sense to fallback here rather than just panic? Is the intention here to be defensive for developers on mobile who might accidentally choose the wrong format for a device? This seems like a dev time bug to me rather than something we'd want to gracefully fail but I'm not sure.

File: crates/bevy_render/src/view/mod.rs [hunk 2]
```diff
@@ -901,7 +893,7 @@ pub fn prepare_view_targets(
                 (a, b, sampled, main_texture)
             });
 
-        let converted_clear_color = clear_color.map(Into::into);
+        let converted_clear_color = clear_color.map(|color| color.into());
```
Comment (tychedelia) [inline crates/bevy_render/src/view/mod.rs:896]: nit: unrelated change.

File: crates/bevy_render/src/texture/mod.rs [hunk 3]
```diff
@@ -170,8 +174,176 @@ pub trait BevyDefault {
...
+            | ViewTargetFormat::Rgb10a2Unorm => wgpu::Features::empty(),
+            ViewTargetFormat::Rb11b10Float => wgpu::Features::RG11B10UFLOAT_RENDERABLE,
+        }
+    }
+
+    /// Unclamped [`ViewTargetFormat`]s are ones whose float values in shader won't be clamped between 0.0 and 1.0.
+    ///
+    /// Bloom and other such effects only work with these formats.
+    pub const fn is_unclamped(&self) -> bool {
+        // BEFORE_MERGE: Heavily bikeshed the name
```
Comment (tychedelia) [inline crates/bevy_render/src/texture/mod.rs:284]: clamping is the language used by the spec so i think this is correct

File: crates/bevy_render/src/camera/camera.rs [hunk 4]
```diff
@@ -508,6 +508,12 @@ impl Default for CameraOutputMode {
     }
 }
 
+pub enum IntermediateTextureFormat {
```
Comment (tychedelia) [inline crates/bevy_render/src/camera/camera.rs:511]: nit: unused

File: crates/bevy_render/src/texture/mod.rs [hunk 5]
```diff
@@ -170,8 +174,176 @@ pub trait BevyDefault {
...
+    // R32Float,
+    // /// Red and green channels. 32 bit float per channel. Float in shader.
+    // Rg32Float,
+    // /// Red, green, blue, and alpha channels. 32 bit float per channel. Float in shader.
+    // Rgba32Float,
+}
+
+impl ViewTargetFormat {
+    pub const DEFAULT: Self = Self::Rgba8UnormSrgb;
+    pub const UNCLAMPED_DEFAULT: Self = Self::Rgba16Float;
```
Comment (tychedelia) [inline crates/bevy_render/src/texture/mod.rs:258]: I don't mind the unclamped name for the function but I think this may be more confusing as a public default for users.

**Review by tychedelia — COMMENTED** [review]
Extremely excellent change, very excited about this! Just some minor nits.

**Comment by IceSentry** [comment]
I still think this is a good direction but this will probably need a significant amount of work to get it working again.
